### PR TITLE
DO-1470 - Add RPC diff and remove absInt function

### PR DESCRIPTION
**Pull request by David-Melamed**
add diff between rpc and reference and remove absInt function to reduce false alarms. This has already been tested and it works properly.

Gatos does not alert when the difference in value is negative:

Screen Shot 2023-07-24 at 15 17 46

### CODE REVIEW

File: core/condition.go [hunk 1]
```diff
@@ -397,18 +397,17 @@ func blockNum(result *Result, first, second string) bool {
...
 }
 
-func absInt(number int64) int64 {
-	if number < 0 {
-		return -number
-	} else {
-		return number
-	}
-}
+// func absInt(number int64) int64 {
```
Comment (0xAleksaOpacic) [inline core/condition.go]: If we are not using this function we can delete it not comment it

**Review by 0xAleksaOpacic — APPROVED** [review]
LFTM!

File: core/condition.go [hunk 2]
```diff
@@ -397,18 +397,17 @@ func blockNum(result *Result, first, second string) bool {
 		return false
 
 	}
-
-	result.AddError(fmt.Sprintf("rpc: %d, reference: %d", firstI, secondI))
-	return !(absInt(secondI-firstI) > deltaThreshold)
+	result.AddError(fmt.Sprintf("rpc: %d, reference: %d, diff: %d", firstI, secondI, secondI-firstI))
+	return !(secondI-firstI > deltaThreshold)
```
Comment (heitorPB) [inline core/condition.go]: Please revert the change on this line. We must use an absolute comparison, otherwise it doesn't make sense to have a delta comparison.

See commit 3ca8636e6eb9d1cdec772f5cb92c028a582730ab.

File: core/condition.go (hunk 1 above)
Comment (heitorPB) [reply core/condition.go]: Please un-comment this function, it is required to not have false alarms.

**Review by heitorPB — CHANGES_REQUESTED** [review]

File: core/condition.go (hunk 2 above)
Comment (David-Melamed) [reply core/condition.go]: Why do we need an absolute comparison? If we keep it as it is, it will continue to alert when the public RPC is behind our node.

Comment (heitorPB) [reply core/condition.go]: Here's the situation we are dealing with: we have two nodes of a distributed system. Both are advancing in state, each one in _their own rhythm_.

We want to make sure that at any given point in time, both nodes "agree" to each other. But it is impossible to guarantee that they will always be in perfect sync, so we allow a small difference (i.e. `deltaThreshold` in the code) between them. Whenever we talk about distributed systems, we expect this small difference in state. And it is fine, as long as it is a _small_ difference. It does not matter which node is ahead, that's why we need an absolute comparison.

By removing the `absInt()` function, whenever the `reference` is ahead of our `rpc` by `1` block (regardless of the `deltaThreshold` value), the check will fail and generate an alert.

Distributed systems are very difficult to understand and tame, and even the smaller code changes can have a huge impact in the overall machinery. Ping me if you want to go deeper into this, I can gladly talk about it for hours :)

Comment (David-Melamed) [reply core/condition.go]: The theoretical part is clear to me, but in practice, I'm missing why we want to receive notifications when the public RPC (reference) is behind our node. Note that I tested this change and all scenario tests passed. In addition, it will reduce false alarms.

Comment (heitorPB) [reply core/condition.go]: > I'm missing why we want to receive notifications when the public RPC (reference) is behind our node

Because we need to know when to change the reference, otherwise we will have false positives.

Comment (David-Melamed) [reply core/condition.go]: Ok, for now, I'll just add the diff field, but we'll have to find a way to deal with problematic public RPCs

**Review by heitorPB — APPROVED** [review]
